File: backend/src/telemetry/simulator.ts
/**
 * Realtime telemetry simulator
 * Shape : pulse wave — 1 cycle = CYCLE_TICKS ticks = 2 hours (at 1-min interval)
 * Noise : Gaussian σ=5% + random spikes P=3% + slow sinusoidal drift ±30%
 */

import { WsGateway } from '../websocket/ws.gateway';
import { TelemetryRepository } from '../modules/telemetry/telemetry.repository';
import { AlertService } from '../modules/alerts/alert.service';
import { TelemetryData } from '../types';

const CYCLE_TICKS  = 120; // 1 full pulse = 120 ticks = 2 hours at 1-min/tick
const TRANS_TICKS  = 5;   // smooth edge over ±5 ticks

// ─── Pulse wave + layered noise ───────────────────────────────────────────────
function pulse(
  threshold:   number,
  tick:        number,
  phaseOffset: number,
  precision:   number,
  dutyCycle  = 0.45,
): number {
  const amplitude = threshold * 0.1;
  const period    = CYCLE_TICKS;
  const phase     = ((tick + phaseOffset) % period + period) % period;
  const highEnd   = dutyCycle * period;

  // ── Pulse shape ──────────────────────────────────────────────────────────
  let base: number;
  if (phase < highEnd - TRANS_TICKS) {
    base = amplitude;
  } else if (phase < highEnd + TRANS_TICKS) {
    const t = (phase - (highEnd - TRANS_TICKS)) / (2 * TRANS_TICKS);
    base = amplitude * (1 - t) + (-amplitude) * t;                       // falling edge
  } else if (phase < period - TRANS_TICKS) {
    base = -amplitude;
  } else {
    const t = (phase - (period - TRANS_TICKS)) / TRANS_TICKS;
    base = -amplitude * (1 - t) + amplitude * t;                         // rising edge
  }

  // ── Noise layer 1: Gaussian (σ = 5% of amplitude) ────────────────────────
  const u1    = Math.random() + 1e-10;
  const u2    = Math.random();
  const gauss = Math.sqrt(-2 * Math.log(u1)) * Math.cos(2 * Math.PI * u2);
  const gaussNoise = gauss * 0.05 * amplitude;

  // ── Noise layer 2: Random spike (P=3%, magnitude ±35% of amplitude) ──────
  const spikeNoise = Math.random() < 0.03
    ? (Math.random() - 0.5) * 0.70 * amplitude
    : 0;

  // ── Noise layer 3: Slow sinusoidal drift (period = 8 cycles = 16 hours) ──
  const drift = (amplitude * 0.30) * Math.sin((2 * Math.PI * tick) / (8 * CYCLE_TICKS));

  const value = threshold + base + gaussNoise + spikeNoise + drift;
  return +Math.max(threshold * 0.82, Math.min(threshold * 1.18, value)).toFixed(precision);
}

// ─── Cumulative counters for vision camera ────────────────────────────────────
const vcState = { inspected: 0, passed: 0, failed: 0 };

// ─── Generators ──────────────────────────────────────────────────────────────
function generateCheckweigher(tick: number): TelemetryData {
  const rejects = Math.max(0, Math.round(pulse(1.5, tick, 10, 0, 0.30)));
  return {
    weight:      pulse(500,  tick,  0, 2, 0.45),
    speed:       pulse(60,   tick,  5, 1, 0.50),
    throughput:  pulse(60,   tick,  5, 1, 0.50),
    rejects,
    status_code: rejects > 0 ? 1 : 0,
  };
}

function generateTemperatureSensor(tick: number): TelemetryData {
  return {
    temp:      pulse(22,  tick,  0, 2, 0.55),
    humidity:  pulse(55,  tick, 20, 1, 0.48),
    dew_point: pulse(11,  tick, 10, 2, 0.52),
  };
}

function generateConveyor(tick: number): TelemetryData {
  return {
    speed:     pulse(1000, tick,  0, 1, 0.45),
    load:      pulse(45,   tick, 30, 1, 0.50),
    rpm:       pulse(750,  tick, 15, 0, 0.45),
    vibration: pulse(5,    tick,  8, 2, 0.40),
  };
}

function generateVisionCamera(tick: number): TelemetryData {
  const defect_rate = pulse(1,  tick,  0, 3, 0.35);
  const confidence  = pulse(97, tick, 25, 1, 0.60);
  const newInspected = Math.floor(Math.random() * 3) + 1;
  vcState.inspected += newInspected;
  const newFailed    = Math.random() < defect_rate / 100 ? 1 : 0;
  vcState.failed    += newFailed;
  vcState.passed    += newInspected - newFailed;
  return {
    defect_rate,
    confidence,
    inspected: vcState.inspected,
    passed:    vcState.passed,
    failed:    vcState.failed,
  };
}

// ─── Simulator ────────────────────────────────────────────────────────────────
interface SimulatedMachine {
  id: string;
  name: string;
  type: 'checkweigher' | 'temperature_sensor' | 'conveyor' | 'vision_camera';
  generate: (tick: number) => TelemetryData;
}

export class TelemetrySimulator {
  private gateway:       WsGateway;
  private telemetryRepo: TelemetryRepository;
  private alertService:  AlertService;
  private machines:      SimulatedMachine[] = [];
  private timer:         NodeJS.Timeout | null = null;
  private intervalMs:    number;
  private persistEvery:  number;
  private tickCount = 0;

  constructor(gateway: WsGateway, intervalMs = 60_000, persistEvery = 1) {
    this.gateway       = gateway;
    this.telemetryRepo = new TelemetryRepository();
    this.alertService  = new AlertService();
    this.intervalMs    = intervalMs;
    this.persistEvery  = persistEvery;
  }

  configureMachines(machines: Array<{ id: string; name: string; type: string }>) {
    this.machines = machines.map(m => {
      let generate: (tick: number) => TelemetryData;
      switch (m.type) {
        case 'checkweigher':       generate = generateCheckweigher;      break;
        case 'temperature_sensor': generate = generateTemperatureSensor; break;
        case 'conveyor':           generate = generateConveyor;          break;
        case 'vision_camera':      generate = generateVisionCamera;      break;
        default:                   generate = () => ({ value: 0 });
      }
      return { id: m.id, name: m.name, type: m.type as SimulatedMachine['type'], generate };
    });
    console.log(`🤖 Simulator configured for ${this.machines.length} machines (pulse wave, 2-hr cycle, layered noise)`);
  }

  start() {
    if (this.timer) return;
    this.timer = setInterval(async () => {
      this.tickCount++;
      await this.tick(this.tickCount % this.persistEvery === 0);
    }, this.intervalMs);
    console.log(`⚡ Simulator started — ${this.intervalMs / 1000}s/tick, 1 cycle = ${CYCLE_TICKS} ticks (2 hrs)`);
  }

  stop() {
    if (this.timer) { clearInterval(this.timer); this.timer = null; }
    console.log('⏹  Simulator stopped');
  }

  private async tick(persist: boolean): Promise<void> {
    const timestamp = new Date();
    await Promise.allSettled(this.machines.map(m => this.processMachine(m, timestamp, persist)));
  }

  private async processMachine(machine: SimulatedMachine, timestamp: Date, persist: boolean): Promise<void> {
    try {
      const data = machine.generate(this.tickCount);
      this.gateway.broadcastTelemetry({
        machineId: machine.id, machineName: machine.name,
        timestamp: timestamp.toISOString(), data,
      });
      if (persist) await this.telemetryRepo.ingest(machine.id, data, timestamp);
      const triggered = await this.alertService.evaluateTelemetry(machine.id, data);
      for (const alert of triggered) {
        this.gateway.broadcastAlert({
          alertId: alert.alertId, alertName: alert.alertName,
          machineId: machine.id, machineName: machine.name,
          field: alert.field, value: alert.value, threshold: alert.threshold,
          condition: alert.condition, severity: alert.severity as 'info' | 'warning' | 'critical',
          message: alert.message, timestamp: timestamp.toISOString(),
        });
      }
    } catch (err) {
      console.error(`[Simulator] ${machine.name}:`, (err as Error).message);
    }
  }
}

File: backend/internal/simulator/simulator.go
package simulator

import (
	"context"
	"fmt"
	"iot-dashboard/internal/modules/alerts"
	ws "iot-dashboard/internal/websocket"
	"math"
	"math/rand"
	"time"
)

// Simulator replicates the TypeScript TelemetrySimulator:
// - Pulse wave with random duty cycle each cycle (120 ticks = 2 hours)
// - Noise layers: Gaussian, spikes, sinusoidal drift
// - Broadcasts via WebSocket, evaluates alert rules.

const (
	cycleTicks = 120 // 2 hours at 1-min ticks
	transTicks = 5   // smooth transition between high/low
)

type MachineConfig struct {
	ID   string
	Name string
	Type string
}

type fieldState struct {
	key       string
	threshold float64
	tick      int
	duty      float64 // 0–1
	driftPhase float64
	pwmPhase  float64
}

type Simulator struct {
	gateway      *ws.Gateway
	alertSvc     *alerts.Service
	machines     []MachineConfig
	tickInterval time.Duration
	stop         chan struct{}
}

func NewSimulator(g *ws.Gateway, tickIntervalMs int) *Simulator {
	return &Simulator{
		gateway:      g,
		alertSvc:     alerts.NewService(),
		tickInterval: time.Duration(tickIntervalMs) * time.Millisecond,
		stop:         make(chan struct{}),
	}
}

func (s *Simulator) ConfigureMachines(machines []MachineConfig) {
	s.machines = machines
}

func (s *Simulator) Start() {
	fmt.Printf("🤖 Simulator started — %d machines, tick every %v\n", len(s.machines), s.tickInterval)
	go s.run()
}

func (s *Simulator) Stop() {
	close(s.stop)
}

func (s *Simulator) run() {
	ticker := time.NewTicker(s.tickInterval)
	defer ticker.Stop()

	// Init field states per machine
	states := make(map[string][]fieldState)
	for _, m := range s.machines {
		states[m.ID] = initFields(m.Type)
	}

	tick := 0
	for {
		select {
		case <-s.stop:
			return
		case <-ticker.C:
			tick++
			ctx := context.Background()

			for _, m := range s.machines {
				fields := states[m.ID]
				data := make(map[string]interface{})

				for i := range fields {
					f := &fields[i]
					val := f.nextValue(tick)
					data[f.key] = val
				}
				states[m.ID] = fields

				// Broadcast via WebSocket
				s.gateway.BroadcastTelemetry(ws.TelemetryPayload{
					MachineID:   m.ID,
					MachineName: m.Name,
					Timestamp:   time.Now().UTC().Format(time.RFC3339),
					Data:        data,
				})

				// Evaluate alert rules
				triggered, err := s.alertSvc.EvaluateTelemetry(ctx, m.ID, data)
				if err == nil {
					for _, t := range triggered {
						s.gateway.BroadcastAlert(ws.AlertPayload{
							AlertID:   t.AlertID,
							AlertName: t.AlertName,
							MachineID: m.ID,
							MachineName: m.Name,
							Field:     t.Field,
							Value:     t.Value,
							Threshold: t.Threshold,
							Condition: t.Condition,
							Severity:  t.Severity,
							Message:   t.Message,
							Timestamp: time.Now().UTC().Format(time.RFC3339),
						})
					}
				}
			}
		}
	}
}

// initFields returns field states for each machine type.
func initFields(machineType string) []fieldState {
	switch machineType {
	case "checkweigher":
		return []fieldState{
			{key: "weight", threshold: 500, duty: randDuty()},
			{key: "speed", threshold: 60, duty: randDuty()},
			{key: "throughput", threshold: 1200, duty: randDuty()},
			{key: "rejects", threshold: 5, duty: randDuty()},
		}
	case "temperature_sensor":
		return []fieldState{
			{key: "temp", threshold: 22, duty: randDuty()},
			{key: "humidity", threshold: 55, duty: randDuty()},
			{key: "dew_point", threshold: 12, duty: randDuty()},
		}
	case "conveyor":
		return []fieldState{
			{key: "speed", threshold: 1.5, duty: randDuty()},
			{key: "load", threshold: 80, duty: randDuty()},
			{key: "rpm", threshold: 750, duty: randDuty()},
			{key: "vibration", threshold: 0.5, duty: randDuty()},
		}
	case "vision_camera":
		return []fieldState{
			{key: "defect_rate", threshold: 2.0, duty: randDuty()},
			{key: "confidence", threshold: 98.5, duty: randDuty()},
		}
	default:
		return []fieldState{{key: "value", threshold: 100, duty: randDuty()}}
	}
}

func (f *fieldState) nextValue(tick int) float64 {
	// Re-randomise duty cycle at start of each cycle
	cyclePos := tick % cycleTicks
	if cyclePos == 0 {
		f.duty = randDuty()
	}

	// Pulse wave: HIGH for duty*cycleTicks ticks, LOW for remainder
	highTicks := int(f.duty * float64(cycleTicks))
	var base float64
	if cyclePos < transTicks {
		// Rising transition from previous state
		prev := f.threshold * 0.90
		next := f.threshold * 1.10
		if cyclePos < highTicks {
			base = lerp(prev, next, float64(cyclePos)/float64(transTicks))
		} else {
			base = lerp(next, prev, float64(cyclePos)/float64(transTicks))
		}
	} else if cyclePos < highTicks {
		base = f.threshold * 1.10 // HIGH plateau
	} else if cyclePos < highTicks+transTicks {
		base = lerp(f.threshold*1.10, f.threshold*0.90, float64(cyclePos-highTicks)/float64(transTicks))
	} else {
		base = f.threshold * 0.90 // LOW plateau
	}

	// Noise layer 1: Gaussian (σ = 5% of threshold)
	gauss := randNorm() * 0.05 * f.threshold
	// Noise layer 2: random spikes (P=3%, ±35%)
	spike := 0.0
	if rand.Float64() < 0.03 {
		spike = (rand.Float64()*2 - 1) * 0.35 * f.threshold
	}
	// Noise layer 3: slow sinusoidal drift (16h period, ±30%)
	f.driftPhase += 2 * math.Pi / (16 * 60) // 16h in minutes
	drift := math.Sin(f.driftPhase) * 0.30 * f.threshold * 0.10

	val := base + gauss + spike + drift

	// Hard clamp: ±18% of threshold
	lo := f.threshold * 0.82
	hi := f.threshold * 1.18
	if val < lo {
		val = lo
	}
	if val > hi {
		val = hi
	}
	return math.Round(val*100) / 100
}

func randDuty() float64 {
	// Random duty cycle between 30% and 70%
	return 0.30 + rand.Float64()*0.40
}

func lerp(a, b, t float64) float64 { return a + (b-a)*t }

func randNorm() float64 {
	// Box-Muller approximation
	u1 := rand.Float64()
	u2 := rand.Float64()
	return math.Sqrt(-2*math.Log(u1+1e-9)) * math.Cos(2*math.Pi*u2)
}
